akka-cluster-tools/src/main/scala/akka/cluster/pubsub/DistributedPubSubMediator.scala

```diff
@@ -258,6 +290,47 @@ object DistributedPubSubMediator {
 
     def encName(s: String) = URLEncoder.encode(s, "utf-8")
 
+    def mkKey(ref: ActorRef): String = mkKey(ref.path)
+
+    def mkKey(path: ActorPath): String = path.toStringWithoutAddress
+
+    trait PerGroupingBuffer {
+      private type BufferedMessages = Vector[(Any, ActorRef)]
+
+      private var buffers: Map[String, BufferedMessages] = Map.empty
+
+      private var totalBufferSize = 0
+
+      def bufferOr(limit: Int, grouping: String, message: Any, originalSender: ActorRef, discard: Any ⇒ Unit)(action: ⇒ Unit): Unit = {
+        buffers.get(grouping) match {
+          case None ⇒ action
+          case Some(messages) if totalBufferSize < limit ⇒
+            buffers = buffers.updated(grouping, messages :+ ((message, originalSender)))
+            totalBufferSize += 1
+          case Some(_) ⇒ discard(message)
+        }
+      }
+
+      def recreateAndForwardMessagesIfNeeded(grouping: String, recipient: ⇒ ActorRef): Unit = {
+        buffers.get(grouping).filter(_.nonEmpty).foreach(forwardMessages(_, recipient))
+        buffers -= grouping
+      }
+
+      def forwardMessages(grouping: String, recipient: ActorRef): Unit = {
+        buffers.get(grouping).foreach(forwardMessages(_, recipient))
+        buffers -= grouping
```

Comment:
I'd prefer to update the `totalBufferSize` at the same place as where the `buffers` is updated